### PR TITLE
IPROTO-307 @Proto annotation

### PR DESCRIPTION
**Pull request by tristantarrant**
https://issues.redhat.com/browse/IPROTO-307
https://issues.redhat.com/browse/IPROTO-306

Adds a `@Proto` annotation for simple handling of protobuf entities. Also adds support for Java `record`s

### CODE REVIEW

File: core/src/main/java/org/infinispan/protostream/annotations/Proto.java [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+import java.lang.annotation.Documented;
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+/**
+ * Defines a Protocol Buffers message without having to annotate all fields with {@link ProtoField}.
+ * Use this annotation to quickly generate messages from records or classes with public fields.
+ * Fields must be public and they will be assigned incremental numbers based on the declaration order.
```
Comment (ryanemerson) [inline core/src/main/java/org/infinispan/protostream/annotations/Proto.java:12]: We should detail here that it's possible to overwrite fields with subsequent @ProtoField definitions.

File: processor/src/test/java/org/infinispan/protostream/annotations/impl/processor/tests/ProtoSchemaTest.java [hunk 2]
```diff
@@ -1376,6 +1379,33 @@ public void testGenericMessage() throws Exception {
...
+
+   @Test
+   public void testBareMessage() {
+      SerializationContext ctx = ProtobufUtil.newSerializationContext();
+
+      GeneratedSchema generatedSchema = new TestBareMessageSerializationContextInitializerImpl();
+      generatedSchema.registerSchema(ctx);
+      generatedSchema.registerMarshallers(ctx);
+
+      assertTrue(generatedSchema.getProtoFile().contains("message BareMessage"));
```
Comment (ryanemerson) [inline processor/src/test/java/org/infinispan/protostream/annotations/impl/processor/tests/ProtoSchemaTest.java:1406]: Should we also assert the generated fields contain the correct types?

File: core/src/main/java/org/infinispan/protostream/annotations/impl/ProtoMessageTypeMetadata.java [hunk 3]
```diff
@@ -719,6 +555,257 @@ private void discoverFields(XClass clazz, Set<XClass> examinedClasses, Map<Integ
...
+               XClass javaType = getJavaTypeFromAnnotation(annotation);
+               if (javaType == typeFactory.fromClass(void.class)) {
+                  javaType = isRepeated ? field.determineRepeatedElementType() : field.getType();
+               }
+               if (javaType == typeFactory.fromClass(byte[].class) && protobufType == Type.MESSAGE) {
+                  // MESSAGE is the default and stands for 'undefined', we can override it with a better default
+                  protobufType = Type.BYTES;
+               }
+               if (!javaType.isArray() && !javaType.isPrimitive() && javaType.isAbstract() && !javaType.isEnum()) {
+                  throw new ProtoSchemaBuilderException("The type " + javaType.getCanonicalName() + " of field '" + fieldName + "' of " + clazz.getCanonicalName() + " should not be abstract, while processing " + this.protoSchemaGenerator.generator);
```
Comment (ryanemerson) [inline core/src/main/java/org/infinispan/protostream/annotations/impl/ProtoMessageTypeMetadata.java:612]: We throw the same exception in two places. We should either split the duplicated code into a method, or at least create the exception via the Log.

**Comment by karesti** [comment]
This is making my day

File: processor/src/test/java/org/infinispan/protostream/annotations/impl/processor/tests/ProtoSchemaTest.java (hunk 2 above)
Comment (tristantarrant) [reply processor/src/test/java/org/infinispan/protostream/annotations/impl/processor/tests/ProtoSchemaTest.java:1411]: Done

File: core/src/main/java/org/infinispan/protostream/annotations/impl/ProtoMessageTypeMetadata.java (hunk 3 above)
Comment (tristantarrant) [reply core/src/main/java/org/infinispan/protostream/annotations/impl/ProtoMessageTypeMetadata.java]: I've promoted a few logs to `Log`, including this one.

File: core/src/main/java/org/infinispan/protostream/annotations/Proto.java (hunk 1 above)
Comment (tristantarrant) [reply core/src/main/java/org/infinispan/protostream/annotations/Proto.java:12]: Done

**Review by ryanemerson — APPROVED** [review]

**Pull request merged**